tests/utils_tests/test_datastructures.py

```diff
@@ -115,6 +166,80 @@ def test_getlist_none_empty_values(self):
         self.assertIsNone(x.getlist('a'))
         self.assertEqual(x.getlist('b'), [])
 
+    def test_setitem(self):
+        x = MultiValueDict({'a': [1, 2]})
+        x['a'] = 3
+        self.assertEqual(list(x.lists()), [('a', [3])])
+
+    def test_setdefault(self):
+        x = MultiValueDict({'a': [1, 2]})
+        a = x.setdefault('a', 3)
+        b = x.setdefault('b', 3)
+        self.assertEqual(a, 2)
+        self.assertEqual(b, 3)
+        self.assertEqual(list(x.lists()), [('a', [1, 2]), ('b', [3])])
+
+    def test_update_too_many_args(self):
+        x = MultiValueDict({'a': []})
+        msg = 'update expected at most 1 argument, got 2'
+        with self.assertRaisesMessage(TypeError, msg):
+            x.update(1, 2)
+
+    def test_update_no_args(self):
+        x = MultiValueDict({'a': []})
+        x.update()
+        self.assertEqual(list(x.lists()), [('a', [])])
+
+    def test_update_dict_arg(self):
+        x = MultiValueDict({'a': [1], 'b': [2], 'c': [3]})
+        x.update({'a': 4, 'b': 5})
+        self.assertEqual(list(x.lists()), [('a', [1, 4]), ('b', [2, 5]), ('c', [3])])
+
+    def test_update_multivaluedict_arg(self):
+        x = MultiValueDict({'a': [1], 'b': [2], 'c': [3]})
+        x.update(MultiValueDict({'a': [4], 'b': [5]}))
+        self.assertEqual(list(x.lists()), [('a', [1, 4]), ('b', [2, 5]), ('c', [3])])
+
+    def test_update_kwargs(self):
+        x = MultiValueDict({'a': [1], 'b': [2], 'c': [3]})
+        x.update(a=4, b=5)
+        self.assertEqual(list(x.lists()), [('a', [1, 4]), ('b', [2, 5]), ('c', [3])])
+
+    def test_update_with_empty_iterable(self):
+        for value in ['', b'', (), [], set(), {}]:
+            d = MultiValueDict()
+            d.update(value)
+            self.assertEqual(d, MultiValueDict())
+
+    def test_update_with_iterable_of_pairs(self):
+        for value in [(('a', 1),), [('a', 1)], {('a', 1)}]:
+            d = MultiValueDict()
+            d.update(value)
+            self.assertEqual(d, MultiValueDict({'a': [1]}))
+
+    def test_update_with_non_iterable_raises_typeerror(self):
+        for value in [None, True, False, 123, 123.45]:
+            type_ = type(value)
+            msg = "'%s' object is not iterable" % type_.__name__
+            with self.subTest(type_), self.assertRaisesMessage(TypeError, msg):
+                MultiValueDict().update(value)
+
+    def test_update_with_unspupported_non_empty_iterable_raises(self):
+        # TypeError: cannot convert dictionary update sequence element #0 to a sequence
+        for value in [b'123', b'abc', (1, 2, 3), [1, 2, 3], {1, 2, 3}]:
+            if PY37:
+                msg = 'cannot unpack non-iterable int object'
+            else:
+                msg = "'int' object is not iterable"
```

Comment:
The message changed in Python 3.7, see [bpo32259](https://bugs.python.org/issue32259).

Reply:
@carltongibson  Could we add an autolink with prefix `bpo-` and target `https://bugs.python.org/issue<num>`? It would be handy :sunglasses:

Reply:
TBH I don't much value in testing Python's error messages, `self.assertRaises(TypeError)` should be enough.

Reply:
That's fair. Sometimes I'm unsure whether we should or not, so I erred on the side of doing so. Will remove.